File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
sys.setrecursionlimit(10 ** 6)
import collections

# 구현 방법 : DFS, BFS 모두 구현하고, 방문여부만 0 1로 구분한뒤, 방문했을 시 그 노드 값을 차례로 임의의 배열에 집어넣은 후 이를 차례대로 출력하는 방식 사용

# DFS와 BFS 모두 구현하기

N, M, V = map(int, sys.stdin.readline().split(" "))

# 그래프 선언
graph = [[] for i in range(N + 1)]

for i in range(M):
    a, b = map(int, sys.stdin.readline().split(" "))
    graph[a].append(b)
    graph[b].append(a)

for s in graph:
    s.sort()

# 방문여부와 순서는 BFS와 DFS에 따라 결과가 다르기에 한번씩 재정의
visited = [0 for i in range(N + 1)]
DFS_ans = [] # DFS 정답

# DFS 구현 함수
def DFS(start):
    DFS_ans.append(start)
    visited[start] = 1
    for s in graph[start]:
        if(visited[s]):
            continue
        DFS(s)
DFS(V)

# BFS 나오는 부분 이기에 재정의
visited = [0 for i in range(N + 1)]
BFS_ans = [] # BFS 정답
# BFS 구현 함수

def BFS(start):
    BFS_ans.append(start)
    visited[start] = 1
    q = collections.deque([start])
    while q:
        start = q.popleft()
        for s in graph[start]:
            if(visited[s] == 0):
                visited[s] = 1
                BFS_ans.append(s)
                q.append(s)

BFS(V)

# 정답 출력
print(" ".join(map(str, DFS_ans))) # DFS 정답
print(" ".join(map(str, BFS_ans))) # BFS 정답

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.util.*;
import java.io.*;

public class Main{

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    private static boolean[] visited;
    private static int[] answer;
    private static int seq = 0;

    private static void DFS(int start) {
        visited[start] = true;
        seq += 1;
        answer[seq] = start;
        for(int i = 0 ; i < graph.get(start).size() ; i++) {
            int newDot = graph.get(start).get(i);
            if(visited[newDot] == false) {
                DFS(newDot);
            }
        }
    }

    private static void BFS(int start) {
        visited[start] = true;
        seq += 1;
        answer[seq] = start;
        Deque<Integer> d = new LinkedList<>();
        d.add(start);
        while(!d.isEmpty()) {
            int dot = d.poll();
            for(int i = 0; i < graph.get(dot).size() ; i++) {
                int newDot = graph.get(dot).get(i);
                if(visited[newDot] == false) {
                    seq += 1;
                    visited[newDot] = true;
                    d.add(newDot);
                    answer[seq] = newDot;
                }
            }
        }

    }


    public static void main(String[] args) throws IOException {
        
        int N, M, V;

        st = new StringTokenizer(br.readLine(), " ");

        N = Integer.parseInt(st.nextToken());

        M = Integer.parseInt(st.nextToken());

        V = Integer.parseInt(st.nextToken());

        for(int i = 0 ; i <= N ; i++) {
            graph.add(new ArrayList<>());
        }

        for(int i = 0 ; i < M ; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            graph.get(start).add(end);
            graph.get(end).add(start);
        }

        for(int i = 0 ; i < graph.size() ; i++) {
            Collections.sort(graph.get(i));
        }

        //  DFS 함수 실행을 위한 초가화
        visited = new boolean[N + 1];
        answer = new int[N + 1];
        DFS(V);

        StringBuilder sb = new StringBuilder();

        for(int i = 1 ; i <= N ; i++) {
            if(answer[i] != 0) {
                sb.append(Integer.toString(answer[i]));
                sb.append(" ");
            }
        }

        bw.write(sb.toString());
        bw.write("\n");

        //  BFS 함수 실행을 위한 초기화
        seq = 0;
        visited = new boolean[N + 1];
        answer = new int[N + 1];
        BFS(V);

        sb = new StringBuilder();

        for(int i = 1 ; i <= N ; i++) {
            if(answer[i] != 0) {
                sb.append(Integer.toString(answer[i]));
                sb.append(" ");
            }
        }

        bw.write(sb.toString());

        bw.flush();

        br.close();
        bw.close();
    }
}